apps/findmydevice/js/findmydevice.js

```diff
@@ -258,70 +298,38 @@ var FindMyDevice = {
-        this.endHighPriority('clientLogic');
-        return;
+        return Promise.resolve();
       }
 
-      navigator.mozId.request();
+      assertionPromise = this._requestAssertion();
     } else {
       // We don't need to send assertions on re-registrations, just rely on
       // our client ID and HAWK signature to authenticate us. This works even
       // if we're logged out of FxA, so use null for the assertion.
-      this._continueRegistration(null);
-    }
-  },
-
-  _onLogin: function fmd_on_login(assertion) {
-    DUMP('logged in to FxA');
-
-    // XXX(ggp) When initializing, we'll process FxA's automatic invocation
-    // of onlogin/onlogout before any IAC message, so use these calls to
-    // initialize _loggedIn. However, when FMD is already running, there's
-    // no guaranteee that onlogin/onlogout will fire before the IAC handler,
-    // so we have dedicated IAC messages (IAC_API_WAKEUP_REASON_LOGIN and
-    // IAC_API_WAKEUP_REASON_LOGOUT) to cover login state changes, and we
-    // use the handlers for these messages to update _loggedIn.
-    this._loggedIn = true;
-
-    if (this._refreshingClientID) {
-      DUMP('resuming client id refresh');
-      this._fetchClientID(assertion);
+      assertionPromise = null;
     }
 
-    if (!this._enabled || !this._registering) {
-      return;
-    }
-
-    // We are in the middle of a registration, so continue here
-    this._continueRegistration(assertion);
-  },
-
-  _continueRegistration: function fmd_continue_registration(assertion) {
-    var self = this;
-
-    if (assertion == null && this._state == null) {
-      // this shouldn't happen
-      throw new Error('Trying to register with no assertion and no state!');
-    }
-
-    var pushRequest = navigator.push.register();
-    pushRequest.onsuccess = function fmd_push_handler() {
-      DUMP('findmydevice received push endpoint!');
-
-      var endpoint = pushRequest.result;
-      if (self._enabled) {
-        self._requestRegistration(assertion, endpoint);
-      }
-
-      self._registering = false;
-    };
-
-    pushRequest.onerror = function fmd_push_error_handler() {
-      DUMP('findmydevice push request failed!');
-
-      self._registering = false;
-      self._countRegistrationRetry();
-    };
+    var pushPromise = navigator.push.register();
+    return Promise.all([assertionPromise, pushPromise]).then(values => {
+      DUMP('got assertion and push endpoint, requesting registration');
+      var assertion = values[0], endpoint = values[1];
+      return this._requestRegistration(assertion, endpoint).then(response => {
+        DUMP('findmydevice successfully registered: ', response);
+        return new Promise((resolve, reject) => {
+          this._registering = false;
+          asyncStorage.setItem('findmydevice-state', response, () => {
+            resolve(this._registeredHelper.set(true));
+          });
+        });
+      }, err => {
+        this._registering = false;
```

Comment:
One remark that apply everywhere in the code, I think it might be nicer if we encapsulated this kind of stuff instead of directly manipulating variables ?